doc/designs/fast-param-find.md

```diff
@@ -0,0 +1,80 @@
+Format:
+
+```
+    %% FUNCTION-BASE-NAME
+
+    prologue
+
+    % PARAM-NAME type [flags]
+    code to process param
+
+    % PARAM-NAME type [flags]
+    code to process param
+
+    %
+
+    epilogue
+```
+
+- the % characters must be start of line
+
+- FUNCTION-BASE-NAME is the name the functions will derive from e.g. "kdf_scrypt" to this will be appended _[gs]et[_ctx]_params
+
+- type is the standard OSSL_PARAM types with additions for basic C types
+
+- Flags include:
+  - alias (uses same name as another entry but different type)
+  - multiple (multiple instances are allowed - not sure if necessary)
+
+
+The idea is that the gettable and get functions will be simultaneously generated
+along with a perfect hash to decode the legal parameters quickly.
+
+
+
+A trivial example:
+```
+    generate_ctx_get(p{
+        %% kdf_scrypt get ctx
+        # no prologue
+        %% KDF_PARAM_SIZE size_t
+            if ((p = OSSL_PARAM_locate(params, OSSL_KDF_PARAM_SIZE)) != NULL)
```

Comment:
Is the idea that OSSL_PARAM_locate would be auto-substituted here? I think I'd probably have some codegen-specific symbol, maybe `@locate` or something.

Reply:
An oversight on my part.  We can't replace locate calls IMO.  They are still useful for repeated params -- but not a performance problem since they always come in sequentially (currently) and a strcmp is minimal effort.